kernel/src/sched.rs

```diff
@@ -900,4 +732,236 @@ impl Kernel {
+                }
+                process.set_syscall_return_value(rval);
+            }
+            Syscall::Yield { which, address } => {
+                if config::CONFIG.trace_syscalls {
+                    debug!("[{:?}] yield. which: {}", process.appid(), which);
+                }
+                if which > (YieldCall::Wait as usize) {
+                    // Only 0 and 1 are valid
+                    return;
+                }
+                let wait = which == (YieldCall::Wait as usize);
+                // If this is a yield-no-wait AND there are no pending
+                // tasks, then return immediately. Otherwise, go into the
+                // yielded state and execute tasks now or when they arrive.
+                let return_now = !wait && !process.has_tasks();
+                if return_now {
+                    // Set the "did I trigger callbacks" flag to be 0,
+                    // return immediately. If address is invalid does
+                    // nothing.
+                    process.set_byte(address, 0);
+                } else {
+                    // There are already enqueued callbacks to execute or
+                    // we should wait for them: handle in the next loop
+                    // iteration and set the "did I trigger callbacks" flag
+                    // to be 1. If address is invalid does nothing.
+                    process.set_byte(address, 1);
+                    process.set_yielded_state();
+                }
+            }
+            Syscall::Subscribe {
+                driver_number,
+                subdriver_number,
+                callback_ptr,
+                appdata,
+            } => {
+                // A callback is identified as a tuple of
+                // the driver number and the subdriver
+                // number.
+                let callback_id = CallbackId {
+                    driver_num: driver_number,
+                    subscribe_num: subdriver_number,
+                };
+                // Only one callback should exist per tuple.
+                // To ensure that there are no pending
+                // callbacks with the same identifier but
+                // with the old function pointer, we clear
+                // them now.
+                process.remove_pending_callbacks(callback_id);
+
+                let ptr = NonNull::new(callback_ptr);
+                let callback = ptr.map_or(Callback::default(), |ptr| {
+                    Callback::new(process.appid(), callback_id, appdata, ptr.cast())
+                });
+                let rval = platform.with_driver(driver_number, |driver| match driver {
+                    Some(d) => {
+                        let res = d.subscribe(subdriver_number, callback, process.appid());
+                        match res {
+                            Ok(newcb) => newcb.into_subscribe_success(),
+                            Err((newcb, err)) => newcb.into_subscribe_failure(err),
```

Comment:
The old callback is returned, correct?
```suggestion
                            Ok(oldcb) => oldcb.into_subscribe_success(),
                            Err((oldcb, err)) => oldcb.into_subscribe_failure(err),
```

Reply:
Not necessarily. Ok should return the previous upcall, while Err will return the passed one.